tests/validator/validator.cpp

```diff
@@ -608,20 +605,67 @@ TEST(Validator, invalidMathMLVariables)
+    const std::vector<std::string> expectedErrors = {
+        "CellML identifiers must contain one or more basic Latin alphabetic characters.",
+        "Model does not have a valid name attribute.",
+        "CellML identifiers must contain one or more basic Latin alphabetic characters.",
+        "Component does not have a valid name attribute.",
+        "MathML ci element has the child text 'B' which does not correspond with any variable names present in component ''.",
+        "W3C MathML DTD error: No declaration for attribute units of element ci.",
+    };
+    libcellml::Validator v;
+    libcellml::ModelPtr m = std::make_shared<libcellml::Model>();
+    libcellml::ComponentPtr c = std::make_shared<libcellml::Component>();
+
+    c->setMath(math);
+    m->addComponent(c);
+
+    v.validateModel(m);
+    EXPECT_EQ_ERRORS(expectedErrors, v);
+}
+
+TEST(Validator, invalidMathmlCellMLNsOnNode)
+{
+    const std::string math =
+        "<math  xmlns:cellml=\"http://www.cellml.org/cellml/2.0#\" xmlns=\"http://www.w3.org/1998/Math/MathML\"><apply><cellml:bvar><ci cellml:units=\"dimensionless\">B</ci></cellml:bvar></apply></math>";
+    const std::vector<std::string> expectedErrors = {
+        "CellML identifiers must contain one or more basic Latin alphabetic characters.",
+        "Model does not have a valid name attribute.",
+        "CellML identifiers must contain one or more basic Latin alphabetic characters.",
+        "Component does not have a valid name attribute.",
+        "Math has a 'bvar' element that is not a supported MathML element.",
+        "MathML ci element has the child text 'B' which does not correspond with any variable names present in component ''.",
+        "W3C MathML DTD error: No declaration for attribute units of element ci.",
+    };
+    libcellml::Validator v;
+    libcellml::ModelPtr m = std::make_shared<libcellml::Model>();
+    libcellml::ComponentPtr c = std::make_shared<libcellml::Component>();
+
+    c->setMath(math);
+    m->addComponent(c);
+
+    v.validateModel(m);
+    EXPECT_EQ_ERRORS(expectedErrors, v);
+}
+
 TEST(Validator, invalidMathMLCiAndCnElementsWithCellMLUnits)
 {
     const std::string math =
         "<math xmlns:cellml=\"http://www.cellml.org/cellml/2.0#\" xmlns=\"http://www.w3.org/1998/Math/MathML\">\n"
         "  <apply>\n"
         "    <eq/>\n"
-        "    <cn cellml:units=\"invalid\">oops</cn>\n"
+        "    <cn cellml:units=\"invalid\" cellml:value=\"zero\">oops</cn>\n"
         "    <apply>\n"
         "      <plus/>\n"
         "      <ci>A</ci>\n"
         "      <apply>\n"
         "        <plus/>\n"
         "        <bvar>\n"
-        "          <ci cellml:units=\"dimensionless\" cellml:value=\"zero\">new_bvar</ci>\n"
+        "          <ci cellml:units=\"dimensionless\">new_bvar</ci>\n"
```

Comment:
Also report on ci elements which have units?

Reply:
No we are not doing that the spec. only mentions `cellml:units` on `cn` elements so letting the MathML DTD complain about units on `ci` elements.

Reply:
Now looking below I understand this even less than before ... in line 700 are we saying that W3C needs a declaration of units for ci elements and that it's not there (but it is) or that it's not allowed? I am confused ...